packages/playwright-core/src/server/trace/recorder/tracing.ts

```diff
@@ -559,11 +559,17 @@ function createActionLogTraceEvent(metadata: CallMetadata, message: string): tra
 function createAfterActionTraceEvent(metadata: CallMetadata): trace.AfterActionTraceEvent | null {
   if (metadata.internal || metadata.method.startsWith('tracing'))
     return null;
+  let error = metadata.error?.error;
+  // Library mode special case for expect errors which are return values, not exceptions.
```

Comment:
Perhaps a separate field in CallMetadata, e.g. `fakeError`? I feel uneasy hardcoding expect here. I'd rather do the hack in `Frame.expect`, not even in `FrameDispatcher.expect`.